contributors/design-proposals/node/limit-node-object-self-control.md

```diff
@@ -0,0 +1,164 @@
+# Limiting Node Scope on the Node object
+
+### Author: Mike Danese, (@mikedanese)
+
+## Background
+
+Today the node client has total authority over its own Node object. This ability
+is incredibly useful for the node auto-registration flow. Some examples of
+fields the kubelet self-reports in the early node object are:
+
+1. Labels (provided by kubelet commandline)
+1. Taints (provided by kubelet commandline)
+
+As well as others.
+
+## Problem
+
+While this distributed method of registration is convenient and expedient, it
+has two problems that a centralized approach would not have. Minorly, it makes
+management difficult. Instead of configuring labels and taints in a centralized
+place, we must configure `N` kubelet command lines. More significantly, the
+approach greatly compromises security. Below are two straightforward escalations
+on an initially compromised node that exhibit the attack vector.
+
+### Capturing Dedicated Workloads
+
+Suppose company `foo` needs to run an application that deals with PII on
+dedicated nodes to comply with government regulation. A common mechanism for
+implementing dedicated nodes in Kubernetes today is to set a label or taint
+(e.g. `foo/dedicated=customer-info-app`) on the node and to select these
+dedicated nodes in the workload controller running `customer-info-app`.
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposed Solution
+
+In many environments, we can improve the situation by centralizing reporting of
+sensitive node attributes to a more trusted source and disallowing reporting of
+these attributes from the kubelet.
+
+### Label And Taint Restriction
+
+An operator will configure a whitelist of taints and labels that nodes are
+allowed to set on themselves. This list should include the taints and labels
+that the kubelet is already setting on itself.
+
+Well known taint keys:
+```
+node.cloudprovider.kubernetes.io/uninitialized
```

Comment:
@kubernetes/sig-scheduling-proposals do kubelets ever modify/remove taints from their node objects, or add taints beyond the initial registration step (based on `--register-with-taints`)? it sounds like they don't.

if not, it seems like we could simply allow nodes to register with whatever taints they want (they're allowed to register with no taints today, and opting into more restrictions should be ok), and simply disallow changes during update. That wouldn't disrupting any existing behavior, and shouldn't require any admission configuration.

Reply:
Can we take the same approach with labels?

Reply:
> Can we take the same approach with labels?

We could possibly do something similar, though we can't just let anything through on create. Adding arbitrary labels on registration lets nodes steer workloads to themselves.

One idea I had was to allow nodes to manage whitelisted labels (the well-known label keys mentioned below), and move any other labels a kubelet attempts to set/clear on itself into an annotation. That does several things:
* preserves the label information the kubelet was trying to set
* makes that information observable via a central controller with higher permissions that could promote those to be real labels on the Node object if desired
* avoids giving kubelets direct control over steering workloads to themselves
* avoids disrupting normal registration/heartbeating activities of kubelets
* avoids needing to configure the NodeRestriction admission plugin

opened https://github.com/kubernetes/kubernetes/pull/64055 as a POC